### ISSUE TITLE
[QUESTION] Is there any plan to support IntelliJ generated code style for Java?

### ISSUE THREAD

**Issue by fabriziocucci** [diffplug/spotless#200]
The line wrapping options that can be configured with IntelliJ _seem_ to be a little bit more flexible than the one available in Eclipse (e.g. lamdas-related preferences).

The closest thing I've found to format Java sources using an IntelliJ code style is its [command line formatter](https://www.jetbrains.com/help/idea/command-line-formatter.html). Although that still requires having the full binaries of IntelliJ (but I hope to be wrong!).

**Comment by nedtwigg** [comment]
Happy to merge a PR for this, but currently nobody is working on this.

You can use a custom step do this, there's an example of how to do this (and a description of its shortcomings) in #119.

**Comment by fabriziocucci** [comment]
Would it be worth opening a separate issue for this _enhancement_ and closing this _question_?

**Comment by fabriziocucci** [comment]
👍

**Comment by fabriziocucci** [comment]
Commented interesting issue on the IntelliJ issue tracker related to this enhancement:

https://youtrack.jetbrains.com/issue/IDEA-180965

**Comment by leonard84** [comment]
Also: Provide standalone formatter or API https://youtrack.jetbrains.com/issue/IDEABKL-7685

**Comment by nedtwigg** [comment]
If you don't care about using the IntelliJ formatter specifically, but rather just the IntelliJ IDE to autoformat in a way that agrees with Spotless, another path forward would be to make a plugin for IntelliJ that calls out to the [spotless IDE hook](https://github.com/diffplug/spotless/blob/master/plugin-gradle/IDE_HOOK.md).  For example, [here's a plugin for VSCode](https://marketplace.visualstudio.com/items?itemName=richardwillis.vscode-spotless-gradle).

**Comment by anandbhaskaran** [comment]
It would be awesome if this come into reality!

**Comment by lfgcampos** [comment]
Long shot but, any news about it?

**Comment by sbilello** [comment]
If we want to use the intellij java code style with spotless. What's the recommended approach? https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml

**Comment by jbduncan** [comment]
@sbilello Spotless doesn't support the IntelliJ formatter yet. The way to get Google-style Java formatting is to use this instruction:

```gradle
spotless {
  java {
    googleJavaFormat().reflowLongStrings()
  }
}
```

It uses [google-java-format](https://github.com/google/google-java-format), which is a successor to the IntelliJ XML file you linked, so I recommend google-java-format over it. :)

**Comment by ragurney** [comment]
Hey! I've actually been working on this as this is something that affects my day-to-day workflow and have a very basic working version of a Spotless IntelliJ plugin working. 

Here is [the code](https://github.com/ragurney/spotless-intellij-gradle).

I'm waiting on JetBrains to approve the plugin for publishing, so it isn't available on the Marketplace yet, but you can install it in your own IDE via <kbd>Settings/Preferences</kbd> > <kbd>Plugins</kbd> > <kbd>⚙️</kbd> > <kbd>Install plugin from disk...</kbd> using [this zip](https://github.com/ragurney/spotless-intellij-gradle/releases/download/v1.0.0/Spotless.Gradle-1.0.0.zip).

It currently only supports an action "Reformat Code with Spotless" which is found under the <kbd>Code</kbd> menu item, but am planning on adding format on save and possibly linting/contextual fixes soon. Feedback encouraged and welcome.

**Comment by leonard84** [comment]
@ragurney that plugin is basically the inverse of this feature request.

**Comment by nedtwigg** [comment]
Fantastic, thanks so much @ragurney! Looks like a great plugin, I will submit a PR momentarily to better advertise your plugin to Spotless users.

If your goal is to use existing Spotless rules with IntelliJ, then this issue is now solved (I'll be using your plugin!).

Some people want something different, which is to have an `intellij()` rule in Spotless which delegates to the formatter which ships in the default IntelliJ IDE. I'll keep this issue open for users who are waiting on that, but my advice to them would be to use a project which is *specific* to formatting (e.g `googlejavaformat`, or `ktfmt`) in conjunction with [your plugin](https://github.com/ragurney/spotless-intellij-gradle).

**Comment by ragurney** [comment]
Got it, thanks so much @nedtwigg! Thank you for the explanation about the other issue here, that makes sense, and it's very exciting that the plugin might be helpful to other Spotless users! I'm hoping to add further functionality to it soon, and looking forward to any feedback or feature requests you might have after using it.

**Comment by lutovich** [comment]
Great plugin @ragurney! 👍

What do you think about adding support for Maven? Would it make sense to have a single Spotless IntelliJ plugin that supports both Gradle and Maven?

**Comment by ragurney** [comment]
Thanks @lutovich!

 I'm leaning towards it being better to have a separate plugin for Maven just for separation of concerns, but if there's a dependable way to tell if an IntelliJ project is using Maven or Gradle that could work too and would be more DRY. I'll do some sleuthing and see if that's a possibility.

**Comment by melissachang** [comment]
I was able to use [File Watchers](https://www.jetbrains.com/help/idea/using-file-watchers.html) to run `./gradlew spotlessApply` on save. You can run any command. Note - Needs to be configured for every repo.

![image](https://user-images.githubusercontent.com/10929390/157562367-704c9cc2-8752-4579-aa74-f083693d6d62.png)

**Comment by beatngu13** [comment]
@melissachang similar to @ragurney's IntelliJ plugin, this helps to execute Spotless via IntelliJ – which is already great!

However, the issue here is basically about loading the IntelliJ formatter into Spotless. For example:

```gradle
spotless {
  java {
    intellij('173').configFile('intellij-prefs.xml')
  }
}
```

**Comment by melissachang** [comment]
Yeah sorry for going a bit off-topic. It took me a while to figure out how to run spotless-on-save (mentioned as a feature request in [this above comment](https://github.com/diffplug/spotless/issues/200#issuecomment-1003646129)). Currently it's not possible with Intellij formatter. Just documenting a way to do it, in case people want spotless-on-save.

**Comment by tonjohn** [comment]
> Regardless, I think a prerequisite for a maven plugin would be adding an IDE hook similar to the [one available for Spotless Gradle](https://github.com/diffplug/spotless/blob/main/plugin-gradle/IDE_HOOK.md).

Is there any particular reason that Spotless Maven lacks parity with Spotless Gradle regarding the IDE hook? 

Are there extra challenges unique to maven in adding such functionality?

(I'm new to the Java ecosystem so apologies for my ignorance)

**Comment by nedtwigg** [comment]
> Is there any particular reason that Spotless Maven lacks parity with Spotless Gradle

Nope, we just got PR's from Gradle users, that's all.

**Comment by lipiridi** [comment]
@ragurney @lutovich @tonjohn @nedtwigg 
Hi! I've been developed a plugin that supports maven based on @ragurney's plugin (for which special thanks to him) and added some new features.
https://plugins.jetbrains.com/plugin/22455-spotless-applier
https://github.com/lipiridi/spotless-applier

**Comment by IlyasYOY** [comment]
Hello! Any updates on the issue. 

I found docs: [Format files from the command line | IntelliJ IDEA Documentation](https://www.jetbrains.com/help/idea/command-line-formatter.html). 

I use this API myself to format code using neovim: 

```lua
vim.keymap.set("n", "<leader>oC", ":!idea format -s ~/idea.xml -r %<CR>edit<CR>")
```

I'd be glad to try to implement the feature if maintainers think it's possible.

PS. I tested this on Linux & MacOS. There some caveats:
- Some IDEA builds have this formatter builtin AFAIU. The version from Jetbrains Tools Box didn't seem to work.
- MacOS version did formatting asynchronously. This is a problem. Version installed in Manjaro works fine.

Anyway considering all above, I'd gave a try to the feature.

**Comment by nedtwigg** [comment]
A PR for this would be very welcome. [`ForeignExe`](https://github.com/diffplug/spotless/blob/main/lib/src/main/java/com/diffplug/spotless/ForeignExe.java) might be the easiest way to implement it. It's okay if it only works on some platforms, so long as it works for the platform you use. Future PR's can always add support for missing platforms.

**Comment by IlyasYOY** [comment]
Thanks for your hint @nedtwigg 

I [created](https://github.com/IlyasYOY/spotless/blob/main/testlib/src/test/java/com/diffplug/spotless/java/IdeaStepTest.java) the `*Step` class a few days ago. It works fine on my macos machine. I faced a problem testing it on CI, there is no IDEA binary & simple installation from brew might require tweaking. 

I saw your PR https://github.com/diffplug/spotless/pull/672. Maybe I'm missing something but I don't have any good ideas on running IDEA in test on every machine (CI at least). I can come up with a solution but maybe the team has an appropriate solution for the problem.

Contribution guide helped a lot, thanks!

**Comment by nedtwigg** [comment]
Amazing! `IdeaStep` has been implemented, after half a decade in the desert! 

Note the following:

- in PR #2001
- gradle/special-tests.gradle
- testlib/src/main/java/com/diffplug/spotless/tag/GofmtTest.java
- note that all tests have `@GofmtTest` from the tag above

In this way, the tests are always compiled, even though they aren't run.

Of all these "special" tests, the only ones which get run on CI right now are the npm tests. It would be better if they all ran, but it would be worse if the steps didn't exist at all.

The important thing is that if someone wants to develop the step further, they aren't starting from scratch - they have tests, they just need to setup their local machine to have foreign exe in question.

**Comment by IlyasYOY** [comment]
Thanks for the advice! Really appreciate it!

I've just opened a Draft PR. 
https://github.com/diffplug/spotless/pull/2020

**Comment by CC007** [comment]
Any info on getting this to work with maven? Spotless applier is a nice first attempt, but is too slow for my use-case.

**Comment by zlsmith86** [comment]
I am excited about this feature!! I hope it makes it to main sometime!  On a new project i ended up using the eclipse formatter because of the customizability, but most of us use IntelliJ so this would be a huge win!!

**Comment by simschla** [comment]
I'm working on this (based on #2020) because I would love to have it too. Can't make any promises on the timeline, though.

**Comment by jimmyzzxhlh** [comment]
@simschla Thank you! Our team is looking into creating a pre-commit hook using Intellij style, so far only Spotless with Intellij formatter seems to be a promising solution.

**Comment by simschla** [comment]
Resolved in #2535

**Comment by nedtwigg** [comment]
Released in `plugin-gradle 7.1.0` and `plugin-maven 2.45.0`.
